tests/core/test_em_asm_arguments_side_effects.cpp

```diff
@@ -0,0 +1,18 @@
+// Copyright 2019 The Emscripten Authors.  All rights reserved.
+// Emscripten is available under two separate licenses, the MIT license and the
+// University of Illinois/NCSA Open Source License.  Both these licenses can be
+// found in the LICENSE file.
+
+#include <stdio.h>
+#include <emscripten.h>
+
+int main(int argc, char **argv) {
+  int counter_1 = 3, counter_2 = 4;
+  // https://github.com/emscripten-core/emscripten/issues/9030
+  int result = EM_ASM_INT({
+     return $0 * 10 + $1;
+  }, counter_1++, counter_2++);
+  printf("counter_1=3, counter_2=4\n");
```

Comment:
I would say `printf("counter_1=%d, counter_2=%d\n", counter_1, counter_2);`, and call that before as well as after the EM_ASM call. We care here not only about the result value, but also the value of the counters after the fact.

In particular, because the order of evaluation is unspecified, this test (only checking result) could pass with the previous version of the code, if the arguments were evaluated right to left. The test should capture the behavior we care about, namely that the counters are only evaluated once by the EM_ASM macro.